equinox/_module/_flatten.py

```diff
@@ -0,0 +1,234 @@
+    data: {{dynamic_type}},
+) -> T:
+    """Generated unflatten function for {{qualname}}.
+
+    Dynamic fields: {{dynamic_fs}}
+    Static fields: {{static_fs}}
+    Wrapper fields: {_WRAPPER_FIELD_NAMES}
+    """
+    self = object.__new__(module_cls)
+    # Set fields directly by index
+{{setters_dynamic}}
+    if aux[0] is not MISSING:
+        waux = aux[0]
+{textwrap.indent(SET_WRAPPERS, INDENT * 2)}
+{{setters_static}}
+    return self
+'''
+
+NS_BASE = {
+    "object": object,
+    "Any": Any,
+    "tuple": tuple,
+    "Literal": Literal,
+    "MISSING": MISSING,
+}
+
+
+def make_tuple_type(count: int, element_type: str = "Any") -> str:
+    """Generate a tuple type annotation string for a given count of elements."""
+    if count == 0:
+        return "tuple[()]"
+    elif count == 1:
+        return f"tuple[{element_type}]"
+    else:
+        return f"tuple[{', '.join([element_type] * count)}]"
+
+
+def _generate_flatten_functions(cls: type, fields: tuple[dataclasses.Field[Any], ...]):
+    """Generate optimized flatten/unflatten functions for a specific field config."""
+    # Separate dynamic and static fields
+    _dynamic_fs, _static_fs = [], []
+    for f in fields:
+        if f.metadata.get("static", False):
+            _static_fs.append(f.name)
+        else:
+            _dynamic_fs.append(f.name)
+    dynamic_fs, static_fs = tuple(_dynamic_fs), tuple(_static_fs)
+
+    # Extract the generated functions from respective namespaces to
+    # set the proper module reference.
+    module_name = getattr(cls, "__module__", "equinox._module._module")
+
+    # -------------------------------------------
+    # Generate flatten function
+
+    # Directly access dynamic fields by name
+    if dynamic_fs:
+        dynamic_exprs = [f"getattr(obj,{name!r},MISSING)" for name in dynamic_fs]
+        dynamic_vals = f"({', '.join(dynamic_exprs)},)"
+    else:
+        dynamic_vals = "()"
+
+    # For static fields, we need to store their values in aux data
+    static_exprs = [f"getattr(obj,{k!r},MISSING)" for k in static_fs]
+    static_vals = f"{', '.join(static_exprs)}"
+
+    # Build return type annotation
+    dynamic_type = make_tuple_type(len(dynamic_fs))
+    wrapper_type = "Literal[MISSING]|tuple[Any, ...]"
+    static_type = f"tuple[{wrapper_type}, {', '.join(['Any'] * len(static_fs))}]"
+    clsname = cls.__qualname__
+
+    # Generate flatten function code
+    flat_code = FLAT_CODE.format(
+        func_name="flatten",
+        return_annotation=f"tuple[{dynamic_type}, {static_type}]",
+        qualname=clsname,
+        dynamic_fs=dynamic_fs,
+        static_fs=static_fs,
+        dynamic_vals=dynamic_vals,
+        static_vals=static_vals,
+    )
+
+    # make flatten func
+    flat_ns = NS_BASE | {"module_cls": cls}
+    exec(compile(flat_code, FLAT_FUNC_NAME.format(clsname), "exec"), flat_ns)
+    flat_fn = flat_ns["flatten"]
+    flat_fn.__module__ = module_name
+    object.__setattr__(flat_fn, "__source__", flat_code)
```

Comment:
Do I understand correctly that `__source__` is not standardized, and this is just a nice addition?

(I like it!)